src/blocks/net.rs

```diff
@@ -105,49 +309,99 @@ fn convert_speed(speed: u64) -> (f64, &'static str) {
+
+        // Update SSID and IP address every 30s.
+        let now = Instant::now();
+        if now.duration_since(self.last_update).as_secs() > 30 {
+            if let Some(ref mut widget) = self.ssid {
+                let ssid = try!(self.device.ssid());
+                if ssid.is_some() {
+                    let mut truncated = ssid.unwrap();
+                    truncated.truncate(self.max_ssid_width);
+                    widget.set_text(truncated);
+                }
+            }
+            if let Some(ref mut widget) = self.ip_addr {
+                let ip_addr = try!(self.device.ip_addr());
+                if ip_addr.is_some() {
+                    widget.set_text(ip_addr.unwrap());
+                }
+            }
+            self.last_update = now;
+        }
+
+        let current_rx = self.device.rx_bytes()?;
         let update_interval = (self.update_interval.as_secs() as f64) + (self.update_interval.subsec_nanos() as f64 / 1_000_000_000.0);
         let rx_bytes = ((current_rx - self.rx_bytes) as f64 / update_interval) as u64;
         let (rx_speed, rx_unit) = convert_speed(rx_bytes);
         self.rx_bytes = current_rx;
 
-        let current_tx = read_file(&format!("{}tx_bytes", self.device_path))?
-            .parse::<u64>()
-            .block_error("net", "failed to parse tx_bytes")?;
+        let current_tx = self.device.tx_bytes()?;
         let tx_bytes = ((current_tx - self.tx_bytes) as f64 / update_interval) as u64;
         let (tx_speed, tx_unit) = convert_speed(tx_bytes);
         self.tx_bytes = current_tx;
 
-        if self.graph {
+        // Update the graph widgets, if they are enabled.
+        if let Some(ref mut widget) = self.graph_rx {
             self.rx_buff.remove(0);
             self.rx_buff.push(rx_bytes);
-
+            widget.set_values(&self.rx_buff, None, None);
+        }
+        if let Some(ref mut widget) = self.graph_tx {
             self.tx_buff.remove(0);
             self.tx_buff.push(tx_bytes);
-
-            self.graph_tx.set_values(&self.tx_buff, None, None);
-            self.graph_rx.set_values(&self.rx_buff, None, None);
+            widget.set_values(&self.tx_buff, None, None);
         }
 
-        self.output_tx
-            .set_text(format!("{:5.1}{}", tx_speed, tx_unit));
-        self.output_rx
-            .set_text(format!("{:5.1}{}", rx_speed, rx_unit));
+        self.output_tx.set_text(
+            format!("{:5.1}{}", tx_speed, tx_unit),
+        );
+        self.output_rx.set_text(
+            format!("{:5.1}{}", rx_speed, rx_unit),
+        );
         Ok(Some(self.update_interval))
     }
 
     fn view(&self) -> Vec<&I3BarWidget> {
-        if self.graph {
-            return vec![
-                &self.output_tx,
-                &self.graph_tx,
-                &self.output_rx,
-                &self.graph_rx,
-            ];
+        // Since we can't error here, report errors as non-up.
+        let is_up = match self.device.is_up() {
```

Comment:
Careful! The `is_up()` function performs I/O, this should not be done in `view()`. Only use this function to render a presentable version of your widget from current state.  `view()` may be called very often (independently of the update intervals set), so it should only perform simple operations and *no* system calls!

Reply:
Fixed by moving state into the `Net` struct itself.